helpers/getLocalizeResources.js

```diff
@@ -0,0 +1,355 @@
+
+	let promise = blobs.get(resource);
+	if (promise === undefined) {
+		promise = fetchWithCaching(resource);
+		blobs.set(resource, promise);
+	}
+	return promise;
+}
+
+function shouldUseBatchFetch() {
+
+	if (documentLocaleSettings === undefined) {
+		documentLocaleSettings = getDocumentLocaleSettings();
+	}
+
+	if (!documentLocaleSettings.oslo) {
+		return false;
+	}
+
+	// Only batch if we can do client-side caching, otherwise it's worse on each
+	// subsequent page navigation.
+
+	return Boolean(documentLocaleSettings.oslo.batch) && 'CacheStorage' in window;
+}
+
+function shouldUseCollectionFetch() {
+
+	if (documentLocaleSettings === undefined) {
+		documentLocaleSettings = getDocumentLocaleSettings();
+	}
+
+	if (!documentLocaleSettings.oslo) {
+		return false;
+	}
+
+	return Boolean(documentLocaleSettings.oslo.collection);
+}
+
+function setVersion(version) {
+
+	if (documentLocaleSettings === undefined) {
+		documentLocaleSettings = getDocumentLocaleSettings();
+	}
+
+	if (!documentLocaleSettings.oslo) {
+		return;
+	}
+
+	documentLocaleSettings.oslo.version = version;
+}
+
+function getVersion() {
+
+	if (documentLocaleSettings === undefined) {
+		documentLocaleSettings = getDocumentLocaleSettings();
+	}
+
+	const shouldReturnVersion =
+		documentLocaleSettings.oslo &&
+		documentLocaleSettings.oslo.version;
+	if (!shouldReturnVersion) {
+		return null;
+	}
+
+	return documentLocaleSettings.oslo.version;
+}
+
+function shouldFetchOverrides() {
+
+	const isOsloAvailable =
+		shouldUseBatchFetch() ||
+		shouldUseCollectionFetch();
+
+	return isOsloAvailable;
+}
+
+function fetchOverride(formatFunc, fetchFunc) {
+
+	let url, res;
+
+	if (shouldUseBatchFetch()) {
+
+		// If batching is available, pool requests together.
+
+		url = formatFunc(OverrideLanguage);
+		url = new URL(url).pathname;
+
+		res = fetchWithPooling(url);
+
+	} else /* shouldUseCollectionFetch() == true */ {
+
+		// Otherwise, fetch it directly and let the LMS manage the cache.
+
+		url = formatFunc(OverrideLanguage);
+		url = new URL(url).pathname;
+		url = documentLocaleSettings.oslo.collection + url;
+
+		res = Promise.resolve(url);
+
+	}
+
+	res = res.then(fetchFunc);
+	res = res.catch(coalesceToNull);
+
+	return res;
+}
+
+function coalesceToNull() {
+
+	return null;
+}
+
+export function __clearWindowCache() {
+
+	// Used to reset state for tests.
+
+	blobs.clear();
+	cache = undefined;
+	cachePromise = undefined;
+}
+
+export async function getLocalizeResources(
+	possibleLanguages,
+	filterFunc,
+	formatFunc,
+	fetchFunc
+) {
+
+	const promises = [];
+	let supportedLanguage;
+
+	if (shouldFetchOverrides()) {
+
+		const overrides = fetchOverride(formatFunc, fetchFunc);
+		promises.push(overrides);
+	}
+
+	for (const language of possibleLanguages) {
+
+		if (filterFunc(language)) {
+
+			if (supportedLanguage === undefined) {
+				supportedLanguage = language;
+			}
+
+			const translations = fetchFunc(formatFunc(language));
```

Comment:
I wasn't expecting a bunch of fetches to happen... only the "best" language match. Why does it need to fetch every language?

Reply:
This is actually what the model component was doing already. Presumably it does this so that, during development, fallback terms are incorporated into the "best" language match. Given that this is all permanently cached after the first hit, and that it's all fetched in parallel over HTTP/2, I didn't think it prudent to change the behaviour.

Reply:
Hmm, that's still surprising to me. I would have expected for the non-static resources case for the dynamic import model inside a switch statement to only fetch the current document's language at runtime. At least in [our example](https://github.com/BrightspaceUI/core/blob/master/mixins/localize-mixin.md#example-2-dynamically-imported-resources) that's what we intended with the `return` statement inside the loop.

Reply:
hmm, I can change it to work like that then. I found multiple examples of doing it this way:

- https://github.com/BrightspaceHypermediaComponents/activities/blob/fa932e04e528af6fae09fede0026458d13053ff4/components/d2l-activity-editor/localization.js
- https://github.com/BrightspaceHypermediaComponents/activities/blob/fa932e04e528af6fae09fede0026458d13053ff4/components/d2l-activity-collection-editor/localization.js
- https://github.com/Brightspace/discovery-fra/blob/48daf747abb9db71bb602f570dc793ad605fd7e4/src/components/d2l-discover-list/localization.js
- https://github.com/Brightspace/attachment/blob/e44cab1f0cecc55dd93acf59212fabc6872c0bd3/helpers/localization.js

But I see now that there are plenty that do it the way you describe, so I'll change it to only fetch the first match.

Reply:
Wow yeah that was definitely not the intent when we made the mixin. I think it makes sense to just fetch the one though right? Like otherwise we'll be seeing ~12X the number of requests -- albeit cached, but browser caches definitely aren't infinite so evictions of valid things might happen.

Reply:
Yeah, I guess I picked a suboptimal component to base this off of (FACE) and then started retconning why it was written like that 😄

Reply:
Retconning... good word.